confluence/provider/client.py

```diff
@@ -133,30 +147,69 @@ def search_pages(self, query):
 
         return response.json().get("results", [])
 
-    def fetch_pages(self, pages):
+    def fetch_pages(self, pages, access_token: str | None = None):
         self._start_session()
-        results = self.loop.run_until_complete(self._gather(pages))
+        results = self.loop.run_until_complete(self._gather(pages, access_token))
         self._close_session_and_loop()
 
         return results
 
-    def search(self, query):
-        pages = self.search_pages(query)
+    def search(self, query, access_token=None):
+        pages = self.search_pages(query, access_token)
 
-        return [page for page in self.fetch_pages(pages) if page is not None]
+        return [
+            page for page in self.fetch_pages(pages, access_token) if page is not None
+        ]
+
+    def _get_headers(self, access_token: str | None = None) -> dict[str, str]:
+        headers = {}
+
+        if access_token:
+            headers["Authorization"] = f"Bearer {access_token}"
+        else:
+            credentials = f"{self.service_user}:{self.service_api_token}"
+            credentials_encoded = base64.b64encode(credentials.encode()).decode("ascii")
+            headers["Authorization"] = f"Basic {credentials_encoded}"
+
+        return headers
+
+    def _get_base_url(self, access_token: str | None = None):
+        if not access_token:
+            return self.service_base_url
+
+        if access_token in self.org_ids:
+            return (
+                f"https://api.atlassian.com/ex/confluence/{self.org_ids[access_token]}"
+            )
+
+        response = requests.get(
+            "https://api.atlassian.com/oauth/token/accessible-resources",
+            headers=self._get_headers(access_token),
+        )
+
+        if response.status_code != 200:
+            logger.error("Error determining Confluence base URL")
+            return
+
+        accessible_resources = response.json()
+
+        if not accessible_resources:
+            logger.error("No resources available to user")
+            return
+
+        org_id = accessible_resources[0]["id"]
+        self.org_ids[access_token] = org_id
+
+        return f"https://api.atlassian.com/ex/confluence/{org_id}"
 
 
 def get_client():
     global client
     if client is None:
-        assert (
-            url := app.config.get("PRODUCT_URL")
-        ), "CONFLUENCE_PRODUCT_URL must be set"
-        assert (user := app.config.get("USER")), "CONFLUENCE_USER must be set"
-        assert (
-            api_token := app.config.get("API_TOKEN")
-        ), "CONFLUENCE_API_TOKEN must be set"
+        product_url = app.config.get("PRODUCT_URL")
+        user = app.config.get("USER")
+        api_token = app.config.get("API_TOKEN")
         search_limit = app.config.get("SEARCH_LIMIT", 10)
-        client = ConfluenceClient(url, user, api_token, search_limit)
+        client = ConfluenceClient(product_url, user, api_token, search_limit)
```

Comment:
@walterbm-cohere I agree, I did not love it either. In an earlier iteration before creating the PR, I had an `auth_type` argument in the `ConfluenceClient` constructor that took `AUTH_TYPE_BASIC` or `AUTH_TYPE_OAUTH` constants as values. I didn't love that either and removed it, and did it based on whether `access_token` was None or not instead. 

I didn't realize until later that there is always a bearer token sent, with both OAuth and Service Auth methods in Cohere Dashboard. When I realized that, deciding whether it was running in OAuth or Service Auth mode based on the access token made less sense.

I did consider whether it should be two separate clients prior to creating the PR, and I also tried to add better validation code to make sure `user`, `api_token` and `product_url` were set when running in Service Auth mode. The issue I ran into however is that there was not an explicit flag anywhere to indicate which mode it should be running in.

If we do it like that, I think I should add a config var to more explicitly state which mode it should start up in, rather than having it support both at the same time, like I was doing here.

I don't think there is an actual security risk with the current code, because usually it would only be set up one way or the other. But I will look at changing it to have separate clients to make it a bit cleaner.

Reply:
I have pushed a change to break it out into into separate `ServiceAuthConfluenceClient` and `OAuthConfluenceClient` classes.